LeaderBankNotifier

apfitzge opened the pull request:
#### Problem
Currently there's not a great way to wait for a new leader slot to start or end, without busy waiting. Often a leader has multiple (4) back-to-back slots, and will end up with txs hitting `PohRecorderError:MaxHeightReached` on transactions during the transition between bank N and bank N+1. Instead of just running a fast loop for the next slot to begin, we can use a `Condvar` and notify other threads when the slot begins/ends.


#### Summary of Changes

- New struct `LeaderBankNotifier`
    - Locked state-machine for slot transitions and notifies on change to `InProgress` or `StandBy`
    - Also stores weak reference to the most recent leader bank  
- Set `InProgress` state when we set the working bank in PohRecorder.
- Set `StandBy` state when the working bank is cleared in PohRecorder.

This also enables us to have less busy threads in BankingStage - probably not going to rework existing BankingStage, but plan to use the functionality in scheduler "worker" threads.

An example of a slot ending notification being useful is in #30396 

Fixes #
<!-- OPTIONAL: Feature Gate Issue: # -->
<!-- Don't forget to add the "feature-gate" label -->

apfitzge commented:
@ryoqun requesting an early review on this. You had a similar mechanism in your branch, which I had adapated for my own uses - using condvar to wake threads up when they have a valid bank to commit to.

LMK what you want/need for this so we can start using a common struct for it.

apfitzge commented:
Thinking that an issue with the current impl is that when we are no longer the leader, we'd end up waiting forever on txs waiting for the bank to commit into.

Probably need to allow some short timeout for the `wait_for_in_progress` as well - maybe wait up to 50ms before we just stop processing txs in the execution thread(s).

ryoqun commented:
> @ryoqun requesting an early review on this. You had a similar mechanism in your branch, which I had adapated for my own uses - using condvar to wake threads up when they have a valid bank to commit to.
> 
> LMK what you want/need for this so we can start using a common struct for it.

hehe, yeah, i had similar struct called `CommitStatus` in my branch. Actually, I'm still not decided that parking worker threads is the best idea during the gap of bank freezing... alternatively, signal retry to scheduler thread, and let the scheduler thread process incoming txes in search of more higher prio tx while waiting for new tpu bank.

That said, the impl is itself looks good at quick glance. As for name, I'd prefer something which alludes blocking nature here. (so, i don't like `CommitStatus` too...)...

ryoqun commented:
this is just fyi, but, as for `TransactionRecorder`, it's synchronization arrangement is too inefficient. namely, one shot channle isn't needed. I'm planning to fix it using this upstream change: https://github.com/crossbeam-rs/crossbeam/pull/959

apfitzge commented:
> this is just fyi, but, as for TransactionRecorder, it's synchronization arrangement is too inefficient. namely, one shot channle isn't needed. I'm planning to fix it using this upstream change: https://github.com/crossbeam-rs/crossbeam/pull/959

Could you elaborate what you mean here? AFAICT `TransactionRecorder` doesn't use a oneshot right now as its' got a loop to receive potentially multiple messages. 100% agree its' inefficient though.

Wasn't entirely clear what you meant in terms of changes with that crossbeam PR - is your desired synchronization in one of your branches?

ryoqun commented:
> > this is just fyi, but, as for TransactionRecorder, it's synchronization arrangement is too inefficient. namely, one shot channle isn't needed. I'm planning to fix it using this upstream change: [crossbeam-rs/crossbeam#959](https://github.com/crossbeam-rs/crossbeam/pull/959)
> 
> Could you elaborate what you mean here? AFAICT `TransactionRecorder` doesn't use a oneshot right now as its' got a loop to receive potentially multiple messages. 100% agree its' inefficient though.

here, it's creating `unbounded()` per a batch. (i.e. per transaction for my unified scheduler.)
https://github.com/solana-labs/solana/blob/65cd55261ac6e91149692deec194871701307287/poh/src/poh_recorder.rs#L212-L224

ryoqun commented:
> Wasn't entirely clear what you meant in terms of changes with that crossbeam PR - is your desired synchronization in one of your branches?

nope, it's only in mind mind and @behzadnouri 's (he was faster than me to report to crossbeam: https://github.com/crossbeam-rs/crossbeam/issues/861 for this optimization opportunity) ... lol 

anyway, it's straightforward: we can remove `.recv_timeout` entirely and FUTEX_WAKE in `.send` in the quoted code in my previous comment.

ryoqun commented:
> The diff coverage is `98.8%`.

btw, diff coverage is quite good. :100:

apfitzge commented:
> > The diff coverage is `98.8%`.
> 
> btw, diff coverage is quite good. 💯

I missed testing the timeout case for both of the functions. Added tests which should bump this to 100%.

ryoqun commented:
> > > The diff coverage is `98.8%`.
> > 
> > 
> > btw, diff coverage is quite good. 100
> 
> I missed testing the timeout case for both of the functions. Added tests which should bump this to 100% - and actually found a bug where even in timeout `wait_for_in_progress` still returned a bank!

see? codecov has some value. :)

<img width="50%" src="https://user-images.githubusercontent.com/117807/226331585-74f7f670-0e64-410e-ad0b-07afc1e4cba6.png"></img>

ryoqun commented on poh/src/poh_recorder.rs:
```diff
@@ -140,25 +140,24 @@ pub struct RecordTransactionsSummary {
     pub is_exited: Arc<AtomicBool>,
-}
-
-impl Clone for TransactionRecorder {
```
Come to think of it, this manual `Clone` impl doen't make sense...

ryoqun commented on poh/src/poh_recorder.rs:
```diff
@@ -308,13 +308,15 @@ pub struct PohRecorder {
     ticks_from_record: u64,
     last_metric: Instant,
     record_sender: Sender<Record>,
+    pub leader_bank_status: Arc<LeaderBankStatus>,
```
nit: i like `pub fn new_leader_bank_status(&self) -> Arc<_>` over `pub`-ing a field.

apfitzge replied:
Done: 0da112f2df0e6048bbb2e4aa6552fff76db98f98

ryoqun commented on poh/src/poh_recorder.rs:
```diff
@@ -308,13 +308,15 @@ pub struct PohRecorder {
 
 impl PohRecorder {
     fn clear_bank(&mut self) {
         if let Some(working_bank) = self.working_bank.take() {
             let bank = working_bank.bank;
+            self.leader_bank_status.set_completed(bank.slot());
```
i'd rewrite this a bit to signify these 3 lines (esp, `.take()`-then-immediately-`.set_completed()` relationship) are strictly coupled semantically:

```rust
if let Some(WorkingBank { bank, start, .. }) = self.working_bank.take() {
  self.leader_bank_status.set_completed(bank.slot());
  ...
  datapoint_info!(
      "leader-slot-start-to-cleared-elapsed-ms",
      ("elapsed", start.elapsed().as_millis(), i64),
   ...  
}
```

apfitzge replied:
Makes sense - done: 5995d31d70345e80551af84e7117047d25af66f0

ryoqun commented on core/src/replay_stage.rs:
```diff
@@ -1802,6 +1806,7 @@ impl ReplayStage {
             banking_tracer.hash_event(parent.slot(), &parent.last_blockhash(), &parent.hash());
 
             let tpu_bank = bank_forks.write().unwrap().insert(tpu_bank);
+            leader_bank_status.set_in_progress(&tpu_bank);
```
once we land #30746 , we can plumb and skip calling this when not used. :)

ryoqun commented on poh/src/poh_recorder.rs:
```diff
@@ -964,6 +970,7 @@ impl PohRecorder {
                 ticks_from_record: 0,
                 last_metric: Instant::now(),
                 record_sender,
+                leader_bank_status: Arc::new(LeaderBankStatus::default()),
```
nit: `Arc::default()` or `LeaderBankStatus::default().into()`

apfitzge replied:
Done: a2344ce98a35e4ca9ae1533d55a41c40589409c6

ryoqun commented on runtime/src/leader_bank_status.rs:
```diff
@@ -0,0 +1,167 @@
+    condvar: Condvar,
+}
+
+/// Leader status state machine for the validator:
+/// [Unininitialized] -> [InProgress] -> [Completed] --|
+///                          ^-------------------------|
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+enum Status {
+    /// Initial state, no bank, but also not completed yet
+    #[default]
+    Uninitialized,
+    /// Slot is in progress as leader
+    InProgress,
+    /// PoH has reached the end of the slot, and the next bank as leader is not available yet
+    Completed,
+}
```
how about removing `Uninitialized`? fewer state, less bugs, fewer bits:

```
enum Status {
  // the leader bank isn't available yet (just after initialized or emptied after poh-completed bank)
  StandBy,
  // poh-initiated bank is available
  InProgress,
}
```

apfitzge replied:
Yeah that makes sense - it seems fine since right now we only ever wait for a *new* notification of completed instead of just being completed/standby.

Done: bea9edccc539d3cc70b78890eab8c47618bbb0e5

ryoqun commented on runtime/src/leader_bank_status.rs:
```diff
@@ -0,0 +1,167 @@
+    /// Initial state, no bank, but also not completed yet
+    #[default]
+    Uninitialized,
+    /// Slot is in progress as leader
```
nit: this comment says same thing as the code.

ryoqun commented on runtime/src/leader_bank_status.rs:
```diff
@@ -0,0 +1,167 @@
+};
+
+/// Tracks leader status of the validator node and notifies when:
+///     1. A leader slot begins
+///     2. A leader slot completes
```
this is a bit vague (maybe in purpose, right?). admittedly, this is in `solana-runtime`, so some generality can be desired in some cases. but, i think it helps reader in this case to be explicit for intended use case like this:

```
/// Tracks leader bank among bank_forks and notifies when:
///     1. A leader bank initiates (=poh-initiated)
///     2. A leader bank completes (=poh-completed)
```

apfitzge replied:
Updated here: bea9edccc539d3cc70b78890eab8c47618bbb0e5

ryoqun commented on runtime/src/leader_bank_status.rs:
```diff
@@ -0,0 +1,167 @@
+///     1. A leader slot begins
+///     2. A leader slot completes
+#[derive(Debug, Default)]
+pub struct LeaderBankStatus {
```
naming: `LeaderBank{Notifier/Tracker/Provider}`?

apfitzge replied:
I like `Notifier` or `Tracker`. Going with `Notifier` since that's the main mechanism of interest

apfitzge replied:
Done: 1b378262622cd58ede1bf0f1695aa320043e0b55 + f1a927cd8aeb2b50658987c1b908d56f4b593e79

ryoqun commented on runtime/src/leader_bank_status.rs:
```diff
@@ -0,0 +1,167 @@
+    /// Wait for next notification for a completed leader slot.
+    /// Returns `None` if the timeout is reached
+    pub fn wait_for_next_completed(&self, mut timeout: Duration) -> Option<Slot> {
+        loop {
```
why different looping construct is used than `.wait_for_in_progress()`?

apfitzge replied:
The naming of the functions is perhaps not clear enough on why these are different.

`wait_for_in_progress` either immediately returns if the status is already `InProgress`, or waits for it to become `InProgress`.

`wait_for_next_completed` (emphasis on **_next_**) explicitly waits for the next notification, regardless of current status.

This is due to my different use-cases for the notifications. I used `wait_for_in_progress` to either get or wait shortly for a bank to process transactions on. I used `wait_for_next_completed` to report metrics at the end of a slot.

ryoqun replied:
oh i see.. how about `get_or_wait_for_in_progress` / `wait_for_complete`? imo, behavior difference is best described by qualifying verbs, not nouns.

apfitzge replied:
done: 438b7bd0ab739ca185ff2d59bd94abc4450da50b

ryoqun commented on runtime/src/leader_bank_status.rs:
```diff
@@ -0,0 +1,167 @@
+
+    /// Wait for next notification for a completed leader slot.
+    /// Returns `None` if the timeout is reached
+    pub fn wait_for_next_completed(&self, mut timeout: Duration) -> Option<Slot> {
```
nit: if `mut `, `remaining_timeout` can help readers.

apfitzge replied:
Done: 1c625ae1325a0ba8ded4dbf87971bb3ae81acf11

ryoqun commented on runtime/src/leader_bank_status.rs:
```diff
@@ -0,0 +1,167 @@
+                matches!(*status, Status::Uninitialized | Status::Completed)
+            })
+            .unwrap();
+
```
ugh, `_status` is `MutexGuard`... I take back the above comment...

ryoqun commented on runtime/src/leader_bank_status.rs:
```diff
@@ -0,0 +1,167 @@
+#[derive(Debug, Default)]
+pub struct LeaderBankStatus {
+    /// Current state of the system
+    status: Mutex<Status>,
+    /// Weak reference to the current bank
+    bank: RwLock<Option<(Slot, Weak<Bank>)>>,
```
why these are separeted? what's wrong with below?:

```rust
slot_and_bank_with_status: Mutex<(Status, Slot, Weak<Bank>)>
```

apfitzge replied:
A lapse in judgement 😦 

Separated an internal state consisting of slot, weak_bank, status: 5cd27833c37c0ef56508a66f62a7781beed9eef3

ryoqun commented on runtime/src/leader_bank_status.rs:
```diff
@@ -0,0 +1,167 @@
+    /// If the status is `InProgress`, immediately return a weak reference to the bank.
+    /// Otherwise, wait up to the `timeout` for the status to become `InProgress`.
+    /// Returns `None` if the timeout is reached.
+    pub fn wait_for_in_progress(&self, timeout: Duration) -> Option<Weak<Bank>> {
```
`Option`-ing `Weak` looks a bit odd. I think `Weak::new()` can be used to flatten this return type? https://doc.rust-lang.org/std/sync/struct.Weak.html#method.new

ryoqun replied:
> ..., so maybe we can just use `new` as you suggested since there's no reason to distinguish. What do you think?

yeah, let's do that.

apfitzge replied:
done: 15858f01955267825ad7392648b53c6f85d01d95

apfitzge commented on poh/src/poh_recorder.rs:
```diff
@@ -422,7 +424,11 @@ impl PohRecorder {
     }
 
     pub fn recorder(&self) -> TransactionRecorder {
```
Separate PR #30838 - function wasn't introduced in this PR

ryoqun commented on runtime/src/lib.rs:
```diff
@@ -45,6 +45,7 @@ pub mod in_mem_accounts_index;
 pub mod inline_spl_associated_token_account;
 pub mod inline_spl_token;
 pub mod inline_spl_token_2022;
+pub mod leader_bank_notifier;
```
btw, why `LeaderBankNotifier` is added in `solana-runtime`? i think `solana-poh` is better suited, or am i missing something?

apfitzge replied:
No you're 100% right.
My initial approach was to put this on the `Bank`, but for various reasons that didn't work as nicely. And I just never re-evaluated the module location.

apfitzge replied:
done: 66bf1e861acdb0f7e29704c83ff30cd5ec5d2041

ryoqun commented on runtime/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,161 @@
+            let (state, result) = self.condvar.wait_timeout(state, remaining_timeout).unwrap();
+            if result.timed_out() {
+                return None;
+            }
+
+            if matches!(state.status, Status::StandBy) {
```
nit: joint with `else if`?

apfitzge replied:
done: 3899f59ec2c92081cb7b0968ded3bbae63e53ff8

ryoqun commented on runtime/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,161 @@
+    /// if the status was not already `InProgress`.
+    pub fn set_in_progress(&self, bank: &Arc<Bank>) {
+        let mut state = self.state.lock().unwrap();
+        if matches!(state.status, Status::InProgress) {
```
i'd add `assert_eq!(state.slot, bank.slot());` just inside at the top of this if body.

moreover, is it necessary to make this fn idempotent? it's only called only once from replay_stage, which should guarantee idempotent isn't needed.

so, I'd remove remove this `if` altogether and replace it with `assert_matches!(state.status, Status::StandBy);` in that way, human reasoning can be simplified and serious logic error at caller side will be caught. imo, idempotent fns are handy sometimes, but it's a kind of variant to swallowing errors (obviously code smell).

apfitzge replied:
removed the if altogether: 8c1b5e2e584fc91e40a1fc770a224823d7499602

ryoqun commented on runtime/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,161 @@
+    ///     2. the slot is higher than the current slot (sanity check).
+    pub fn set_completed(&self, slot: Slot) {
+        let mut state = self.state.lock().unwrap();
+        if matches!(state.status, Status::StandBy) || slot < state.slot {
```
contrasting with https://github.com/solana-labs/solana/pull/30395/files#r1144237575, i understand idempotent is desired for this fn. that's because multiple banking thread can call `set_completed` anytime, right?

if that's correct, how about commenting this `if` to justify to exist as such?

apfitzge replied:
done as part of this commit: 52632ee4d579872be1957c45086f3825af561c64

ryoqun commented on runtime/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,161 @@
+    ///     1. the status was not already `StandBy` and
+    ///     2. the slot is higher than the current slot (sanity check).
+    pub fn set_completed(&self, slot: Slot) {
+        let mut state = self.state.lock().unwrap();
```
also, `slot < state.slot` could turn out to be problematic. i think there's a possibility where leader_slot+1 is created later leader_slot+3 for example, in th event of excessive forking and poh-reset.

considering, set_in_progress is single threaded by definition with replay stage thread and poh thread, i'd rather add `assert_eq!(self.state.slot, slot);` just before the following `if` for strictest sanity check (i.e. before idempotent bail-out). i know, i know, panic are bad. but imo, better than debugging a race condition on mainnet-beta, I'd say.

apfitzge replied:
done: 52632ee4d579872be1957c45086f3825af561c64

ryoqun commented on runtime/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,161 @@
+            }
+
+            if matches!(state.status, Status::StandBy) {
+                return Some(state.slot);
```
i think `Some(0)` or any stale `Some(N)` could wrongly be returned from here due to spurious wakeup of condvars...

in short, it's not guranteed that only once wakeup is triggered for state transitions. this thread could be waken up multiple time while state.status == Status::StandBy still.

so, how about putting `let slot = ...;` outside this `loop`  for bookkeeping to guarntee progression is actually cranked?

apfitzge replied:
We, unfortunately, cannot check for slot changes because it may initially be called while already `InProgress` (same slot as it will be when we call `set_completed`)

I've taken an alternative approach where I explicitly check for state change from `InProgress` -> `StandBy`, waiting for `InProgress` if necessary.

apfitzge replied:
^ commit for above alternative approach with explicit `InProgress` -> `StandBy` transition: 58984d84223e08abc5768fb05794c9c9319ccbcb

ryoqun commented on runtime/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,161 @@
+            }
+
+            remaining_timeout = remaining_timeout.saturating_sub(start.elapsed());
+        }
```
`state` (mutex guard) is immediately dropped here. so, aba problem could be possible.

i mean, indefinite number of cycles of state transition (InProgress => Completed => InProgress => ..) is theoretically possible while this thread has yet to re-enter `.wait_timeout` at the front of next `loop` iteration.

So, I'd suggest putting `let state = ...` outside this loop.

apfitzge replied:
I think I may be missing something. We shadow the original `state` by the `MutexGuard` returned from `wait_timeout`. That should be held until the end of the scope (loop iteration), i.e. through this remaining timeout calculation.

apfitzge replied:
This got somewhat changed due to https://github.com/solana-labs/solana/pull/30395#discussion_r1144267697, but I am still curious if my previous understanding was wrong

ryoqun commented on runtime/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,161 @@
+#[derive(Debug, Default)]
+struct SlotAndBankWithStatus {
+    status: Status,
+    slot: Slot,
```
nit: the `Some(0)` case in https://github.com/solana-labs/solana/pull/30395/files#r1144267697 was possible because of this not `Option<Slot>`. not all of solana code aren't diligent to distinguish block 0 (https://explorer.solana.com/block/0) and None. but I'd prefer Option here a bit.

ryoqun replied:
at least, `Option<Slot>`  can remind us of thinking twice (or thrice, lol) for the uninitialized case silently introduced `#[derive(Default)]`...

apfitzge replied:
100% => 9f41c41c5b516b981c5127ea9553e52f12fe6f88. I even had errors in my tests from thinking the default_for_tests bank was slot 1 not slot 0 🤦

ryoqun reviewed (changes requested):
thanks for working so hard cleaning up the code for my taste... and that lead spotting some subtle bugs.. condvar is hard to write/review code correctly.... sadly, this review round got bigger than expected previously...

apfitzge commented:
Had to rebase due to minor conflict with #30838.

@ryoqun, sorry for the somewhat larger change in 58984d84223e08abc5768fb05794c9c9319ccbcb. I think explicitly waiting for the state transition (by verifying) initial state is the most robust way to get what we want. It also gets rid of the raw `loop` where I really didn't want one 😄 

Appreciate all your feedback!

apfitzge commented:
@ryoqun, hold off on reviewing. Having panics has caused some tests to fail, though I believe they are tests which do not represent situations we can actually enter: trying to record w/o a working bank, trying to record on a non-existent bank slot.

I'm going to work on some split-off PRs tomorrow to address these issues outside this PR (too big already) - certainly the meaning of `MaxHeightReached` is confusingly overloaded.

apfitzge commented on poh/src/poh_recorder.rs:
```diff
@@ -425,6 +424,10 @@ impl PohRecorder {
         TransactionRecorder::new(self.record_sender.clone(), self.is_exited.clone())
     }
 
+    pub fn new_leader_bank_notifier(&self) -> Arc<LeaderBankNotifier> {
```
Currently unused - will be used by banking worker threads.

apfitzge commented on poh/src/poh_recorder.rs:
```diff
@@ -579,6 +582,8 @@ impl PohRecorder {
     }
 
     pub fn set_bank(&mut self, bank: &Arc<Bank>, track_transaction_indexes: bool) {
+        assert!(self.working_bank.is_none());
```
Sanity check here on the logic of `PohRecorder`.

This potentially adds a new panic, but is a bit more direct than if we get the panic in notifier for having 2 in progress banks.

ryoqun commented on poh/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,289 @@
+
+    /// Set the status to `StandBy` and notify any waiting threads if the status was changed.
+    /// Panics if the stored slot does not match the given slot.
+    pub fn set_completed(&self, slot: Slot) {
```
nit: `pub(crate)` ditto with https://github.com/solana-labs/solana/pull/30395/files#r1147575512

apfitzge replied:
ditto: 7c458d90ddc1b4c0e29bdd6e2a08063290e67710

ryoqun commented on poh/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,289 @@
+            .wait_timeout_while(state, timeout, |state| {
+                matches!(state.status, Status::StandBy)
+            })
+            .unwrap();
+
+        (!wait_timeout_result.timed_out()).then_some(state)
```
nit: i think this 3 lines can be chained with `.filter_map()`

apfitzge replied:
I could do a filter_map on the `wait_timeout_while`, but I actually intended for a PoisonError on the mutex to panic with the unwrap. And the wait_timeout_result is unfortunately named, imo, as it is not really a `Result` that we can use the usual chaining methods on.

ryoqun replied:
oh, my bad. yeah, you're correct.

ryoqun commented on poh/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,289 @@
+        let state = self.get_or_wait_for_in_progress_state(remaining_timeout, state);
+
+        let Some(state) = state else { return None; };
+        remaining_timeout = remaining_timeout.saturating_sub(now.elapsed());
```
i learned from this std code as well, but `.checked_sub` is more properly nuanced here i guess: https://doc.rust-lang.org/1.67.1/src/std/sync/condvar.rs.html#460

ryoqun commented on poh/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,289 @@
+
+    // Helper function to get or wait for the `InProgress` status with a given `MutexGuard`.
+    fn get_or_wait_for_in_progress_state<'a>(
+        &self,
```
nit: if this `fn` is called _a helper function_, I'd like not to take `&self`, rather `&'a Condvar`, so that it's very clear the intent to readers and avoids pitfalls like double locking. (admittedly, code is small, so pretty low risk).

Alternatively, if this `fn` is called _a helper method_, `&self` is ok-ish. but, I'd prefer adjusting the code.

ryoqun commented on poh/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,289 @@
+            .wait_timeout_while(state, remaining_timeout, |state| {
+                matches!(state.status, Status::InProgress)
+            })
+            .unwrap();
+
+        if !wait_timeout_result.timed_out() {
+            state.slot
+        } else {
+            None
+        }
```
nit: `.filter_map()` :)

apfitzge replied:
Similarly here, we want to panic on PoisonErrors, and the `wait_timeout_result` is not actually a `Result` type - copied from `condvar.rs`:
```rust
/// A type indicating whether a timed wait on a condition variable returned
/// due to a time out or not.
///
/// It is returned by the [`wait_timeout`] method.
///
/// [`wait_timeout`]: Condvar::wait_timeout
#[derive(Debug, PartialEq, Eq, Copy, Clone)]
#[stable(feature = "wait_timeout", since = "1.5.0")]
pub struct WaitTimeoutResult(bool);
```

The only way we can inspect the "result" is through `timed_out()`, and bools do not implement an `and_then` like options: https://doc.rust-lang.org/std/option/enum.Option.html#method.and_then

Having an if seems cleaner to me than jumping through the hoop of something like:
```rust
    (!wait_timeout_result).then_some(()).and_then(|| state.slot)
```

ryoqun replied:
yeah, my bad; we can't use `.filter_map()` here.

how about this? imo, less clumsy than `then_some(...).and_then(...)`:

```patch
diff --git a/poh/src/leader_bank_notifier.rs b/poh/src/leader_bank_notifier.rs
index 309033ba885..e38d64be344 100644
--- a/poh/src/leader_bank_notifier.rs
+++ b/poh/src/leader_bank_notifier.rs
@@ -96,11 +96,7 @@ impl LeaderBankNotifier {
             })
             .unwrap();
 
-        if !wait_timeout_result.timed_out() {
-            state.slot
-        } else {
-            None
-        }
+        (!wait_timeout_result.timed_out()).then(|| state.slot.expect("some slot whenever completed"))
     }
 
     /// Helper function to get or wait for the `InProgress` status with a given `MutexGuard`.
``` 

or

```patch
@@ -96,11 +96,7 @@ impl LeaderBankNotifier {
             })
             .unwrap();

-        if !wait_timeout_result.timed_out() {
-            state.slot
-        } else {
-            None
-        }
+        (!wait_timeout_result.timed_out()).then(|| state.slot).flatten()
     }

     /// Helper function to get or wait for the `InProgress` status with a given `MutexGuard`.

```

`.flatten()` (and current code) is silently discarding fatal logic error of allowign `state.slot` to be `None` at this point.

apfitzge replied:
Yeah, we probably should verify that the slot is `Some`: 974cf7ee59d910de56ec44ff3a8cbe93de26ac8a

I don't think it's a super great check since after the first `set_in_progress` the slot will never be cleared, but considering multiple threads could be waiting for complete it's not somethng that is easily done 😄

ryoqun commented on poh/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,289 @@
+    // Helper function to get or wait for the `InProgress` status with a given `MutexGuard`.
+    fn get_or_wait_for_in_progress_state<'a>(
+        &self,
+        timeout: Duration,
+        state: MutexGuard<'a, SlotAndBankWithStatus>,
```
nit: arg ordering is an art. but, I'd prefer `state, timeout` (i.e. reverse the current order) to be consistent with the wrapped std fn. (i.e. `.wait_timeout_while(state, timeout)`) for beauty at this fn's call site.

apfitzge replied:
ordering parity is good: 27616ae1ed4c83f0486e160943008a4eb9ad289a

apfitzge commented on poh/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,289 @@
+
+        // If currently `StandBy`, need to wait for `InProgress` to begin.
+        let now = Instant::now();
+        let state = self.get_or_wait_for_in_progress_state(remaining_timeout, state);
+
+        let Some(state) = state else { return None; };
```
Nice catch!

apfitzge commented on poh/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,289 @@
+        let state = self.state.lock().unwrap();
+        self.get_or_wait_for_in_progress_state(timeout, state)
+            .map(|state| state.bank.clone())
+            .unwrap_or_else(Weak::new)
```
done: 770e0e0e2e8f337ffba9ebf7f4c21ffc6f154581

apfitzge commented on poh/src/leader_bank_notifier.rs:
```diff
@@ -0,0 +1,289 @@
+        // Immediately return if the status is `InProgress`
+        if matches!(state.status, Status::InProgress) {
+            return Some(state);
+        }
```
removed here: 56af4b49529033eed2a1c5b19e6324898045e246

thanks for catching it! I made a mistake, and was surprised that condvar didn't check the initial state...turns out I was just being dumb 🎉

apfitzge commented:
@ryoqun, I had to rebase due to some [dependency audit failures](https://discord.com/channels/428295358100013066/439194979856809985/1088665374314201149) unrelated to the change, which screwed up some of the commit links I replied with...

In addition to addressing your comments I added an additional commit 42d778b77b87c21a189138894d7ed49237eb1851 to add appropriate assertions in `set_completed` since it is now only called by the recorder, similar to our previous assertions with `set_in_progress`

ryoqun commented on poh/src/poh_recorder.rs:
```diff
@@ -137,17 +137,24 @@ pub struct RecordTransactionsSummary {
     pub starting_transaction_index: Option<usize>,
 }
 
-#[derive(Clone)]
 pub struct TransactionRecorder {
     // shared by all users of PohRecorder
     pub record_sender: Sender<Record>,
     pub is_exited: Arc<AtomicBool>,
 }
 
+impl Clone for TransactionRecorder {
+    fn clone(&self) -> Self {
+        TransactionRecorder::new(self.record_sender.clone(), self.is_exited.clone())
+    }
+}
+
 impl TransactionRecorder {
     pub fn new(record_sender: Sender<Record>, is_exited: Arc<AtomicBool>) -> Self {
         Self {
+            // shared
             record_sender,
+            // shared
             is_exited,
         }
```
hmm, seems rebase miss... manual `Clone` is revived unexpectedly...

apfitzge replied:
🤦 nice catch. I manually "reverted" my original derive by re-adding these manually.
That's what I get for not using git properly!

apfitzge replied:
Added a single commit to make your re-review simpler, which should do nothing once we squash everything on merge: b679dee9c2bc06a1a8a98c1c4af54c5b9d6b9b3f

ryoqun reviewed (approved):
lgtm with nits.

super hard work here. really appreciate addressing all of my comments seriously.. :) Through that journey, I think the impl pivoting made sense.

At this point, I've thoroughly analyzed the code path and i think all of newly added panic sources are safe.

There's a few of nits still. but i think it's ok to merge as-is.

apfitzge commented:
@ryoqun I pushed those 2 nits...

> unless you're so obsessed with cleanest code possible like me... xD

I think you may have me pegged 😆

ryoqun reviewed (approved):
re-lgtm. thanks for the extra chores.
